File: python/app/webhooks.py
import hashlib
import hmac
import json
import time
from typing import Dict

import httpx


def sign_payload(secret: str, payload: Dict[str, object], timestamp: int) -> str:
    message = f"{timestamp}.{json.dumps(payload, separators=(',', ':'))}"
    digest = hmac.new(secret.encode("utf-8"), message.encode("utf-8"), hashlib.sha256).hexdigest()
    return f"t={timestamp},v1={digest}"


def deliver_webhook(url: str, secret: str, event: Dict[str, object], timeout: float = 5.0) -> None:
    timestamp = int(time.time())
    signature = sign_payload(secret, event, timestamp)
    headers = {
        "Content-Type": "application/json",
        "Payment-Signature": signature,
    }
    response = httpx.post(url, json=event, headers=headers, timeout=timeout)
    response.raise_for_status()

File: typescript/src/common/webhooks.ts
import { createHmac } from "crypto";

export const signPayload = (secret: string, payload: string, timestamp: number) => {
  const message = `${timestamp}.${payload}`;
  const digest = createHmac("sha256", secret).update(message).digest("hex");
  return `t=${timestamp},v1=${digest}`;
};

export const deliverWebhook = async (url: string, secret: string, event: unknown) => {
  const payload = JSON.stringify(event);
  const timestamp = Math.floor(Date.now() / 1000);
  const signature = signPayload(secret, payload, timestamp);
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Payment-Signature": signature
    },
    body: payload
  });
  if (!response.ok) {
    throw new Error(`webhook status ${response.status}`);
  }
};
